posix/mount.c

```diff
@@ -0,0 +1,290 @@
+
+    if (!path || !suffix)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    /* First check whether a device id is set for this thread. */
+    {
+        uint64_t devid;
+
+        if ((devid = oe_get_thread_devid()) != OE_DEVID_NONE)
+        {
+            oe_device_t* device;
+
+            if (!(device = oe_get_device(devid, OE_DEVICE_TYPE_FILESYSTEM)))
+                OE_RAISE_ERRNO(OE_EINVAL);
+
+            /* Use this device. */
+            if (oe_strlcpy(suffix, path, OE_PATH_MAX) >= OE_PATH_MAX)
+                OE_RAISE_ERRNO(OE_ENAMETOOLONG);
+
+            ret = device;
+            goto done;
+        }
+    }
+
+    /* Find the real path (the absolute non-relative path). */
+    if (!oe_realpath(path, realpath))
+        OE_RAISE_ERRNO(oe_errno);
+
+    oe_spin_lock(&_lock);
+    {
+        /* Find the longest binding point that contains this path. */
+        for (size_t i = 0; i < _mount_table_size; i++)
+        {
+            size_t len = oe_strlen(_mount_table[i].path);
+            const char* mpath = _mount_table[i].path;
+
+            if (mpath[0] == '/' && mpath[1] == '\0')
+            {
+                if (len > match_len)
+                {
+                    if (suffix)
+                    {
+                        oe_strlcpy(suffix, realpath, OE_PATH_MAX);
+                    }
+
+                    match_len = len;
+                    ret = _mount_table[i].fs;
+                }
+            }
+            else if (
+                oe_strncmp(mpath, realpath, len) == 0 &&
+                (realpath[len] == '/' || realpath[len] == '\0'))
+            {
+                if (len > match_len)
+                {
+                    if (suffix)
+                    {
+                        oe_strlcpy(suffix, realpath + len, OE_PATH_MAX);
+
+                        if (*suffix == '\0')
+                            oe_strlcpy(suffix, "/", OE_PATH_MAX);
+                    }
+
+                    match_len = len;
+                    ret = _mount_table[i].fs;
+                }
+            }
+        }
+    }
+    oe_spin_unlock(&_lock);
+
+    if (!ret)
+        OE_RAISE_ERRNO_MSG(OE_ENOENT, "path=%s", path);
+
+done:
+    return ret;
+}
+
+int oe_mount(
+    const char* source,
+    const char* target,
+    const char* filesystemtype,
+    unsigned long mountflags,
+    const void* data)
+{
+    int ret = -1;
+    oe_device_t* device = NULL;
+    oe_device_t* new_device = NULL;
+    bool locked = false;
+    int retval = -1;
+
+    OE_UNUSED(data);
+
+    if (!target || !filesystemtype)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    /* Resolve the device for the given filesystemtype. */
+    if (!(device = oe_find_device(filesystemtype, OE_DEVICE_TYPE_FILESYSTEM)))
+        OE_RAISE_ERRNO_MSG(OE_EINVAL, "filesystemtype=%s", filesystemtype);
+
+    /* Be sure the full_target directory exists (if not root). */
+    if (oe_strcmp(target, "/") != 0)
+    {
+        struct oe_stat buf;
+        int retval = -1;
```

Comment:
there are two retvals, one here and the other at 134. Is this one necessary?

Reply:
Fixed. Thanks!